lib/interactive_viewer.ml

```diff
@@ -115,44 +115,163 @@ let toggle_view_mode () =
   | Normal -> Lwd.set view_mode SideBySide
   | SideBySide -> Lwd.set view_mode Normal
 
-let rec split_and_align_hunk hunks mine_acc their_acc =
+(* Somewhere here we are not organizing *)
+(* Adding a type for changes / common / empty *)
+(* Rewrite this function to tag properly  *)
+
+type line = Change of change_origin * string | Common of string | Empty
+and change_origin = Mine | Their
+
+let rec split_and_align_hunk hunks mine_acc their_acc mine_hunk_length
+    their_hunk_length =
+  let rec buffer_changes temp_mine temp_their mine_hunk_length their_hunk_length
+      hunks =
+    match hunks with
+    | [] -> (temp_mine, temp_their, mine_hunk_length, their_hunk_length, hunks)
+    | `Common _ :: _ ->
+        ( temp_mine,
+          temp_their,
+          mine_hunk_length + 1,
+          their_hunk_length + 1,
+          hunks )
+    | `Mine line :: t ->
+        buffer_changes
+          (Change (Mine, line) :: temp_mine)
+          temp_their (mine_hunk_length + 1) their_hunk_length t
+    | `Their line :: t ->
+        buffer_changes temp_mine
+          (Change (Their, line) :: temp_their)
+          mine_hunk_length (their_hunk_length + 1) t
+  in
+
+  let append_balanced temp_mine temp_their mine_hunk_length their_hunk_length =
+    let rec fill_empty lst n =
+      if n <= 0 then lst else fill_empty (Empty :: lst) (n - 1)
+    in
+    let mine_len = List.length temp_mine in
+    let their_len = List.length temp_their in
+    if mine_len > their_len then
```

Comment:
These 3 if cases contain a lot of duplication. Could this be written in a shorter more readable way ?